### ISSUE TITLE
`test_benchdnn...` failures on AArch64 builds.

### ISSUE THREAD

**Issue by nSircombe** [oneapi-src/oneDNN#1108]
# Summary
Two of the standard `test_benchdnn` tests have started failing on AArch64, in both the master and release branches. 
These failures were not caught by the DroneCI runs, which only run the `gtest` subset.
This means that the 2.3 release is currently failing on AArch64, but I'm not yet sure why. I've added details below; do you have any suggestions? do you think this is something that we can resolve for the 2.3 release?

The failing tests are:
```
The following tests FAILED:
	111 - test_benchdnn_gru_ci_cpu (Failed)
	112 - test_benchdnn_lstm_ci_cpu (Failed)
Errors while running CTest
```

These failures were not observed when preparing #1103 as the commit which introduced them was not in our downstream branch at the time.

# Version
The first failing commit appears to be: https://github.com/oneapi-src/oneDNN/commit/79be1a9330ed2e70abbdf5857c8dd0bf39dceb1e

# Environment
* CPU make and model: Arm AArch64, Neoverse-N1
* OS version: Ubuntu 20.04 and RHEL 8.x
* Compiler version: gcc 7, 8, 9, & 10. Clang 9.

# Steps to reproduce
From the cloned oneDNN dir:

```
mkdir build
cd build
cmake -DDNNL_CPU_RUNTIME=OMP -DDNNL_BUILD_FOR_CI=ON ../. && make -j 32
ctest -R test_benchdnn_gru_ci_cpu  --verbose
cd ..
```

# Observed behavior
```
...
110: tests:10240 passed:1280 skipped:8320 mistrusted:0 unimplemented:640 failed:640 listed:0
1/1 Test #110: test_benchdnn_gru_ci_cpu .........***Failed  531.62 sec

0% tests passed, 1 tests failed out of 1
```
There are 640 tests reporting 'UNIMPLEMENTED' which were previously skipped.

# Expected behavior
All benchdnn tests should pass.
Prior to 79be1a9330ed2e70abbdf5857c8dd0bf39dceb1e, `ctest -R test_benchdnn_gru_ci_cpu  --verbose` gave:

```
110: tests:10240 passed:1280 skipped:8960 mistrusted:0 unimplemented:0 failed:0 listed:0
1/1 Test #110: test_benchdnn_gru_ci_cpu .........   Passed  536.62 sec

The following tests passed:
        test_benchdnn_gru_ci_cpu

100% tests passed, 0 tests failed out of 1
```

Work-around - removing the include of `platform.hpp` from `test_isa_common.hpp`...
```
diff --git a/tests/test_isa_common.hpp b/tests/test_isa_common.hpp
index 9599f7d45..c6b5d506f 100644
--- a/tests/test_isa_common.hpp
+++ b/tests/test_isa_common.hpp
@@ -29,7 +29,6 @@
 #include "oneapi/dnnl/dnnl.h"
 #include "oneapi/dnnl/dnnl.hpp"
 
-#include "src/cpu/platform.hpp"
 
 #if DNNL_X64
 #include "src/cpu/x64/cpu_isa_traits.hpp"
```

**Comment by igorsafo** [comment]
Hi @nSircombe ,
I was able to reproduce the issue. All `UNIMPLEMENTED` fails are related to int8 RNN test cases which were added back in April in commit 7951acb. Running benchdnn under GDB shows that RNN implementation `ref_rnn_t` [requires GEMM packing](https://github.com/oneapi-src/oneDNN/blob/57d315ebd773a6a4e43d12a6dc47894ff8d3a326/src/cpu/rnn/rnn_utils.hpp#L730) for int8 which [is not implemented on ARM](https://github.com/oneapi-src/oneDNN/blob/57d315ebd773a6a4e43d12a6dc47894ff8d3a326/src/cpu/gemm/gemm_pack.cpp#L42).

I wasn't able to find any issues related to commit 79be1a9 you mentioned in the description.

Regards,
Igor Safonov

**Comment by nSircombe** [comment]
Hi @igorsafo, thanks for taking a look - it looks like this issue has been lurking for some time, but looking at our CI builds of the upstream master & release branch, it appears it was only exposed when https://github.com/oneapi-src/oneDNN/commit/79be1a9330ed2e70abbdf5857c8dd0bf39dceb1e landed, previously the failing tests were skipped, but I'm not sure why... 

What do you suggest we should do?

**Comment by igorsafo** [comment]
@nSircombe my bad, indeed there is an issue. Adding `#include "src/cpu/platform.hpp"` defines `DNNL_X64 = 0` and [the following check](https://github.com/oneapi-src/oneDNN/blob/57d315ebd773a6a4e43d12a6dc47894ff8d3a326/tests/benchdnn/rnn/rnn.cpp#L788) doesn't work anymore: 
```cpp
#if !defined(DNNL_X64) // it is defined, but equal to 0. Should be #if DNNL_X64 == 0 instead
    // int8 is not supported altogether since RNN relies on packed IGEMM
    // FIXME: this will disable int8 RNN testing if the library is built with
    //        Intel MKL that does have packed IGEMM
    if (prb.is_int8()) {
        res->state = SKIPPED, res->reason = CASE_NOT_SUPPORTED;
        return;
    }
#endif
```

So the good thing there are no issues in the library, however there is an issue in benchdnn, which should be fixed and it is a one line fix. I am not sure this is critical enough to push it into rls-v2.3. Summoning @vpirogov. 


Regards,
Igor Safonov

**Comment by vpirogov** [comment]
Let's fix it in v2.3, no reason to leave AArch64 validation with failures.

**Comment by igorsafo** [comment]
@nSircombe I opened a PR #1109 with a fixe for this issue. Could you please validate it on your side?

**Comment by nSircombe** [comment]
> Let's fix it in v2.3, no reason to leave AArch64 validation with failures.

Thanks @vpirogov that's great.

>  I opened a PR #1109 with a fixe for this issue. Could you please validate it on your side?
Thanks @igorsafo, I'm just trying it now.

**Comment by nSircombe** [comment]
>  I'm just trying it now.

All looks good to me.

**Comment by igorsafo** [comment]
@nSircombe The changes landed in rls-v2.3 (577cc0c). Thanks for heads-up and sorry for inconvenience.

**Comment by nSircombe** [comment]
That's great @igorsafo , thanks for sorting it out so quickly!

Will there be a patch for `master`, or will you consolidate the two branches after the 2.3 release is complete?

**Comment by igorsafo** [comment]
@nSircombe There will be a separate patch for master branch.

**Comment by nSircombe** [comment]
Hi @igorsafo,
With the 2.3 release away now, do you think it would be possible to get https://github.com/oneapi-src/oneDNN/pull/1109 onto the master branch (I'd like to rebase our downstream master which is a few weeks behind now).

**Comment by igorsafo** [comment]
Hi @nSircombe , the changes are already in the internal repository. I will promote them today along with public PRs.

**Comment by nSircombe** [comment]
That's great, thanks @igorsafo!

**Comment by igorsafo** [comment]
@nSircombe The changes are available in master: aa13406e8a284495faaddf5396d9090a1d46f1ca .
Closing the issue since it is addressed in both rls-v2.3 and master branches.

**Comment by nSircombe** [comment]
Thanks @igorsafo, @vpirogov - just to confirm, everything looks good for all our builds now!
